ide/che-ide-gwt-app/src/main/resources/org/eclipse/che/ide/public/loader.html

```diff
@@ -12,11 +12,247 @@
+                        }
+                        resolve(JSON.parse(xhr.responseText));
+                    };
+                });
+            });
+        }
+
+        function setAuthorizationHeader(xhr) {
+            return new Promise((resolve, reject) => {
+                if (window._keycloak && window._keycloak.token) {
+                    window._keycloak.updateToken(5).success(() => {
+                        xhr.setRequestHeader('Authorization', 'Bearer ' + window._keycloak.token);
+                        resolve(xhr);
+                    }).error(() => {
+                        window._keycloak.login();
+                        reject('Failed to refresh token');
+                    });
+                }
+
+                resolve(xhr);
+            });
+        }
+
+        function asyncCheckServiceLink(workspace, redirectUrl) {
+            return new Promise((resolve, reject) => {
+                if (!workspace.runtime) {
+                    reject("Can't check service link: Workspace isn't RUNNING at the moment.");
+                    return;
+                }
+
+                const server = workspace.runtime.links.find(link => redirectUrl.startsWith(link));
+                if (server) {
+                    resolve(server);
+                } else {
+                    reject("Workspace doesn't have a server which matches with URL: " + redirectUrl);
+                }
+            });
+        }
+
+        function asyncGetWsToken(workspace, redirectUrl) {
+            return new Promise((resolve, reject) => {
+                if (workspace.runtime) {
+                    resolve(workspace.runtime.machineToken);
+                } else {
+                    reject("Can't get ws-token: Workspace isn't RUNNING at the moment.");
+                }
+            });
+        }
+
+        function asyncAuthenticate(redirectUrl, token) {
+            const re = new RegExp(/(https?:\/\/[^\/]+?)(?:$|\/).*/),
+                //                  \    /     \     /
+                //                  scheme    host:port
+                url = redirectUrl.replace(re, "$1" + "/authenticate?token=" + encodeURIComponent(token));
+            return new Promise((resolve, reject) => {
+                const request = new XMLHttpRequest();
+                request.open('GET', url);
+                request.send();
+                request.onreadystatechange = () => {
+                    if (request.readyState !== 4) {
+                        return;
+                    }
+                    if (request.status !== 204) {
+                        reject(request.status ? request.statusText : "Unknown error");
+                        return;
+                    }
+                    resolve();
+                };
+            });
+        }
+
+        (function () {
+            const keycloackAuthenticationPromise = new KeycloakLoader().loadKeycloakSettings();
+
+            const workspaceId = getQueryParam('wsId'),
```

Comment:
I have a proposal to rename it to `workspaceId` to make it clearer that it is workspace but not WebSocket or anything else. @skabashnyuk WDYT?

Reply:
+1

Reply:
Fixed workflow described in the issue